Add a nutrition facts label to the custom food dialog

Xebeth opened the pull request:
![image](https://github.com/user-attachments/assets/5b20ebee-2eea-43df-b9e9-d7474023e7e6)

Xebeth commented:
There is probably a better way to do this so I'll leave that one as a draft.

Xebeth commented:
Sorry, I wasn't super clear but I would like your feedback on that one if you've got time to review the code.

jlai commented on src/components/nutrition/food/custom-food.tsx:
```diff
@@ -178,6 +179,24 @@ export function CustomFoodFields() {
+              width={"50%"}
+              servingText={servingText}
+              calories={watch("calories")}
+              caloriesFromFat={watch("nutritionValues.caloriesFromFat")}
```
You could pass in the entire nutritionValues object to the nutrition label as a single property so you don't have to pass each value individually

jlai commented on src/components/nutrition/label/NutritionLabel.tsx:
```diff
@@ -0,0 +1,172 @@
+}));
+
+const Title = styled(Typography)(() => ({
+    fontSize: "21pt",
```
pt is fairly unusual to use on web, as noted in https://css-tricks.com/css-font-size/ . Prefer using px (or rem which is better if the user has custom font sizes, but that gets trickier because you have to do the math: 1rem = 16px)

jlai commented on src/components/nutrition/label/NutritionLabel.tsx:
```diff
@@ -0,0 +1,172 @@
+
+            <SeparatorBar height={"0.5pt"} color={backgroundColor}/>
+
+            <NutrientRow label={"Total Fat"} value={totalFat||0} unit="g" boldLabel={true} recommended={78} />
```
FYI, in React `boldLabel={true}` can also be written as just `boldLabel` for brevity, and omitted for the false case. Spelling it out is fine too, though.

Xebeth commented:
I think I've done pretty much all the features I wanted for that PR (can't speak for the quality of the code, I'm still a React noob)
Options:
![image](https://github.com/user-attachments/assets/d414fa57-dcee-4c35-aeee-8339a2b176b4)
Button in the custom food datagrid:
![image](https://github.com/user-attachments/assets/8c789444-74ea-4013-b023-f2ed1b71a2ee)
Added a toolbar in the food log for the edit button + the nutrition popover:
![image](https://github.com/user-attachments/assets/0f07bafa-eadd-455d-a26d-11bad3be0c92)
If you hold CTRL or ALT, it will show the values for the default serving size instead
![image](https://github.com/user-attachments/assets/5c4eb991-7cf9-4b34-a06b-515e453ecc93)
I put a little warning to say that the data is limited for public food
![image](https://github.com/user-attachments/assets/33166570-44e0-464e-84db-749ccc3b2826)

Xebeth commented:
I've done pretty much all the features I wanted to do for this PR (can't speak for the quality of the code, I'm still a React noob)
Options:
![image](https://github.com/user-attachments/assets/7951334a-8d05-4ea0-8e84-f767029ab459)
Added a button to the data grid of custom foods to show the nutrition facts popover:
![image](https://github.com/user-attachments/assets/90814560-c26f-4a0b-aa9f-d4ce904e521b)
Added a toolbar in the food log to regroup the actions (edit + nutrition facts popover):
![image](https://github.com/user-attachments/assets/e8f0cda0-c6a4-4274-9843-1cb16802e9cd)
If you hold CTRL or ALT when clicking, it will show the values for the default serving size:
![image](https://github.com/user-attachments/assets/6c988229-0934-46a2-a07a-bc0b364b9427)
I added a little warning sign for public foods to explain the nutrition data is limited:
![image](https://github.com/user-attachments/assets/8a8df746-9a45-4014-877d-8803736cb84a)

Xebeth commented:
It seems to work OK on mobile as well.

jlai commented on src/components/nutrition/label/nutrition-popover.tsx:
```diff
@@ -0,0 +1,108 @@
+                                 || context?.foodLog?.nutritionalValues?.carbs || 0 }
+                caloriesFromFat={ food.nutritionalValues?.caloriesFromFat || 0 }
+                saturatedFat={ food.nutritionalValues?.saturatedFat || 0 }
+                dietaryFiber={ food.nutritionalValues?.dietaryFiber
```
Instead of dealing with all these individually, if we use a nutritionalValues object we could do something like:

```ts
nutritionalValues={{
  ...food.nutritionalValues,
  ...context?.foodLog?.nutritionalValues
}}
```

NutritionLabel already has `|| 0` for the values so we don't to do it in two places.

jlai commented on src/components/nutrition/label/separator-bar.tsx:
```diff
@@ -0,0 +1,21 @@
+  )
+}
+
+export default SeparatorBar
```
style: end with semicolon

jlai commented on src/components/nutrition/label/nutrition-popover.tsx:
```diff
@@ -0,0 +1,108 @@
+          ]}
+        >
+          <ClickAwayListener onClickAway={popupState.close}>
+            <Box className="bg-slate-50 dark:bg-slate-900 p-21">
```
Overlays like popovers should use something like `<Paper>` to provide a border and shadow, otherwise it's harder to distinguish from the content below.

Also: `p-21` is not a valid tailwind class (looks like it goes up to 12) so this doesn't do anything

Xebeth replied:
I had a Paper at first but the popper wouldn't close. It seemed to work with a Box instead. Maybe it didn't close for another reason. I'll try Paper again.

jlai commented on src/components/nutrition/label/nutrition-popover.tsx:
```diff
@@ -0,0 +1,108 @@
+                  <WarningOutlined className="float-right p-1" />
+                </Tooltip>
+              )}
+              <Typography variant="subtitle1">{food.name}</Typography>
```
Needs some padding or margin

jlai commented on src/components/nutrition/label/nutrition-label.tsx:
```diff
@@ -0,0 +1,210 @@
+  minWidth?: string,
+  width?: string,
+  digits?: number,
+  recommendedValues?: {
```
Use `NutritionalValues` type?

jlai commented on src/components/nutrition/label/nutrition-label.tsx:
```diff
@@ -0,0 +1,210 @@
+            <SeparatorBar height={"1px"} color={backgroundColor}/>
+
+            <NutrientRow label={"Total Fat"} value={(totalFat || 0) * multiplier} unit="g" boldLabel
+                         recommended={(recValues?.totalFat || 78)} digits={digits} />
```
Instead of defining them here, maybe do something like

```ts
// Somewhere like config/nutrition.ts
const FDA_RECOMMENDED_VALUES = {
  totalFat: 78,
  sodium: 2300
  // ...etc
};
```

```ts
// Merge base recommended values with custom macro goals
const recValues = {
   ...FDA_RECOMMENDED_VALUES,
  ...macroGoals
};
```

Xebeth replied:
Again, trying to avoid coupling

Xebeth commented on src/app/nutrition/dialogs/food-log-rows.tsx:
```diff
@@ -138,37 +144,42 @@ function FoodLogRow({ foodLog }: { foodLog: FoodLogEntry }) {
+            </IconButton>
+            <IconButton size="small" className="text-slate-500" title="Nutrition facts"
+                        onClick={ (event) => setTimeout((target) => {
+                          // without the timeout, the click event interferes with the clickAway event in the popper
```
I'm on brave and without the timeout I get 2 popups or even sometimes 3 if I click them in succession:
![foodlog](https://github.com/user-attachments/assets/46934a43-5cb6-4115-b9a7-89f9c40dd5c0)

Xebeth commented on src/app/settings/foods/custom/page.tsx:
```diff
@@ -38,12 +41,40 @@ const customFoodsColumns: Array<GridColDef<Food>> = [
+
+  return (
+    <GridActionsCellItem
+      onClick={ (event) => setTimeout((target) => {
```
This one is a little different. When I was testing it let me open 2 in succession but if I clicked the first again, it wouldn't open and I had to click again. When I recorded this, it wouldn't even open on the first attempt (needed 2 clicks):
![customfoods+bug](https://github.com/user-attachments/assets/d0521aa6-92ce-46c9-aab5-70627648339d)
whereas with the timeout there is a slight delay but it opens with one click:
![customfoods+timeout](https://github.com/user-attachments/assets/823944f8-39f0-4112-9e24-d77e39295ad0)

jlai replied:
Try https://mui.com/base-ui/react-click-away-listener/#listening-for-leading-events

Xebeth replied:
It works great with that.

Xebeth commented:
Thanks for taking the time to review my code. I'll have a look later.

Xebeth commented on src/components/nutrition/label/nutrition-label.tsx:
```diff
@@ -0,0 +1,210 @@
+    margin: "3px 0 7px"
+}));
+
+function NutritionLabel({ backgroundColor, servingText, calories, caloriesFromFat, totalFat, saturatedFat, transFat,
```
In this particular case, I think it wouldn't be a good idea to add coupling with one of our types. This component could be extracted as its own package.

Xebeth commented:
Updated popper
![image](https://github.com/user-attachments/assets/be208bd0-cc88-4f60-afb2-af2ade942213)

jlai commented on src/app/settings/foods/custom/page.tsx:
```diff
@@ -38,12 +41,40 @@ const customFoodsColumns: Array<GridColDef<Food>> = [
   },
 ];
 
+let popupState: PopupState;
```
This should be passed as a prop, not stored as a global. React can render asynchronously, so storing state in global variables is problematic.

You can move customFoodColumns inside `ManageCustomFoods()` instead of at the top level:

```ts
  const popupState = usePopupState({
    popupId: "show-nutrition-facts-popup",
    variant: "popper",
  });

  const customFoodsColumns: Array<GridColDef<Food>> = [
    { field: "name", headerName: "Food", flex: 2 },
    { field: "brand", headerName: "Brand", flex: 1 },
    {
      field: "accessLevel",
      headerName: "Type",
      valueFormatter: (accessLevel) =>
        accessLevel === "PRIVATE" ? "Custom" : "Public",
    },
    {
      field: "actions",
      type: "actions",
      headerName: "Actions",
      width: 100,
      getActions: ({ id, row: food }: GridRowParams<Food>) => [
        ...(food.accessLevel === "PRIVATE"
          ? [
              <ShowLabelAction key={id} food={food} popupState={popupState} />,
              <EditAction key={id} food={food} />,
            ]
          : []),
      ],
    },
  ];
```

Xebeth replied:
I didn't dare move customFoodsColumns but it makes sense to put both in the component.
